### PR TITLE
[postgresql] Added section on extensions

### PR DESCRIPTION
**Pull request by chhei**
Added a section on enabling Uberspace-provided extensions to PostgreSQL installations with two examples for [PostGIS](http://postgis.net) and the [UUID-OSSP extension](https://www.postgresql.org/docs/12/uuid-ossp.html). Some minor fixes on style and typos.

This is the fix for #1082

### CODE REVIEW

**Comment by brutus** [comment]
@chhei Great work, thanks! 

I'll take a closer look soon, for now just some quick cleanup suggestions:

```diff
diff --git a/source/guide_postgresql.rst b/source/guide_postgresql.rst
index 11a1da4..2e8ddcf 100644
--- a/source/guide_postgresql.rst
+++ b/source/guide_postgresql.rst
@@ -203,10 +203,10 @@ Consider using only unix sockets if possible.
 
 .. warning:: Please replace ``<username>`` with your username!
 
-.. warning:: If you set listen_addresses you might open your postgres installation to the world!
+.. warning:: If you set ``listen_addresses`` you might open your postgres installation to the world!
 
 .. code-block:: postgres
- :emphasize-lines: 14
+ :emphasize-lines: 7,11,14
 
  #------------------------------------------------------------------------------
  # CONNECTIONS AND AUTHENTICATION
@@ -231,10 +231,13 @@ Consider using only unix sockets if possible.
  #bonjour_name = ''                     # defaults to the computer name
                                         # (change requires restart)
 
-.. code-block:: postgres
- :emphasize-lines: 14
+Edit the "Reporting and Logging" section in
+``~/opt/postgresql/data/postgresql.conf`` to enable logging. Consider setting
+the ``log_directory`` to your preferred log file output location (here we use
+``/home/<username>/logs`` where ``username`` is the name of your Uberspace).
 
-Edit the "Reporting and Logging" section in ``~/opt/postgresql/data/postgresql.conf`` to enable logging. Consider setting the ``log_directory`` to your preferred log file output location (here we use ``/home/<username>/logs`` where ``username`` is the name of your Uberspace).
+.. code-block:: postgres
+ :emphasize-lines: 19
 
  #------------------------------------------------------------------------------
  # REPORTING AND LOGGING
@@ -254,7 +257,7 @@ Edit the "Reporting and Logging" section in ``~/opt/postgresql/data/postgresql.c
                                          # (change requires restart)
 
  # These are only used if logging_collector is on:
- log_directory = '/home/<username>/logs'                    # directory where log files are written,
+ log_directory = '/home/<username>/logs' # directory where log files are written,
                                          # can be absolute or relative to PGDATA
  log_filename = 'postgresql-%a.log'      # log file name pattern,
                                          # can include strftime() escapes
```

**Comment by chhei** [comment]
@brutus - thanks for those, implemented, and also moved the "socket" part further up.

```
diff --git a/source/guide_postgresql.rst b/source/guide_postgresql.rst
index 11a1da4..41f3a31 100644
--- a/source/guide_postgresql.rst
+++ b/source/guide_postgresql.rst
@@ -203,10 +203,10 @@ Consider using only unix sockets if possible.
 
 .. warning:: Please replace ``<username>`` with your username!
 
-.. warning:: If you set listen_addresses you might open your postgres installation to the world!
+.. warning:: If you set ``listen_addresses`` you might open your postgres installation to the world!
 
 .. code-block:: postgres
- :emphasize-lines: 14
+ :emphasize-lines: 7,11,14
 
  #------------------------------------------------------------------------------
  # CONNECTIONS AND AUTHENTICATION
@@ -231,11 +231,14 @@ Consider using only unix sockets if possible.
  #bonjour_name = ''                     # defaults to the computer name
                                         # (change requires restart)
 
-.. code-block:: postgres
- :emphasize-lines: 14
+Later you can see the socket in the filesystem by using ``ls -a ~/tmp``. It is listed as ``.s.PGSQL.5432``.
+
 
 Edit the "Reporting and Logging" section in ``~/opt/postgresql/data/postgresql.conf`` to enable logging. Consider setting the ``log_directory`` to your preferred log file output location (here we use ``/home/<username>/logs`` where ``username`` is the name of your Uberspace).
 
+.. code-block:: postgres
+ :emphasize-lines: 19
+
  #------------------------------------------------------------------------------
  # REPORTING AND LOGGING
  #------------------------------------------------------------------------------
@@ -254,7 +257,7 @@ Edit the "Reporting and Logging" section in ``~/opt/postgresql/data/postgresql.c
                                          # (change requires restart)
 
  # These are only used if logging_collector is on:
- log_directory = '/home/<username>/logs'                    # directory where log files are written,
+ log_directory = '/home/<username>/logs' # directory where log files are written,
                                          # can be absolute or relative to PGDATA
  log_filename = 'postgresql-%a.log'      # log file name pattern,
                                          # can include strftime() escapes
@@ -276,11 +279,6 @@ Edit the "Reporting and Logging" section in ``~/opt/postgresql/data/postgresql.c
 
 
 
-Later you can see the socket in the filesystem by using ``ls -a ~/tmp``. It is listed as ``.s.PGSQL.5432``.
-
-
-
-
 Setup Daemon
 ------------
 

```

File: source/guide_postgresql.rst [hunk 1]
```diff
@@ -298,14 +350,96 @@ Create Database
...
+
+.. code-block:: console
+
+ [isabell@stardust ~]$ ls -rtl /usr/pgsql-12/share/extension/postgis--3*.sql
+ [isabell@stardust ~]$ -rw-r--r--. 1 root root 7.9M May 26 15:59 /usr/pgsql-12/share/extension/postgis--3.1.2.sql
+
+
+Once you have convinced yourself that the right PostGIS_ extension is available, simply enter the database which you have just created (`Create Database`_) as administrator and spatially `enable the database <http://postgis.net/docs/postgis_administration.html#create_spatial_db>`_ using the PostGIS extension:
+
+.. warning:: Please replace ``<user>`` with the user name, created earlier, and ``<database name>`` with the name of the database you want to create!
```
Comment (brutus) [inline source/guide_postgresql.rst:389]:
```suggestion
.. warning:: Please replace ``<user>`` and ``<database name>`` with the values you used earlier (when you created the database).
```

File: source/guide_postgresql.rst [hunk 2]
```diff
@@ -298,14 +350,96 @@ Create Database
...
+
+.. code-block:: console
+
+ [isabell@stardust ~]$ psql <database name> -U<user>
```
Comment (brutus) [inline source/guide_postgresql.rst:393]: I thinke we should prefer _long options_:
```suggestion
[isabell@stardust ~]$ psql <database name> --username <user>
```

File: source/guide_postgresql.rst [hunk 3]
```diff
@@ -298,14 +350,96 @@ Create Database
...
+ databaseName=# CREATE EXTENSION IF NOT EXISTS plpgsql;
+ databaseName=# CREATE EXTENSION postgis;
+ databaseName=# CREATE EXTENSION postgis_raster; -- OPTIONAL
+ databaseName=# CREATE EXTENSION postgis_topology; -- OPTIONAL
+
+The ``raster`` and ``topology`` functionality of PostGIS_ are optional. Test whether the extensions have been properly installed and can be found by PostgreSQL using the ``\dx`` - you should get output similar to this:
+
+.. code-block:: console
+
+ [isabell@stardust ~]$ psql <database name> -U<user>
```
Comment (brutus) [inline source/guide_postgresql.rst:403]:
```suggestion
[isabell@stardust ~]$ psql <database name> --username <user>
```

File: source/guide_postgresql.rst [hunk 4]
```diff
@@ -226,9 +231,56 @@ Consider using only unix sockets if possible.
  #bonjour_name = ''                     # defaults to the computer name
                                         # (change requires restart)
 
+.. code-block:: postgres
+ :emphasize-lines: 14
+
+Edit the "Reporting and Logging" section in ``~/opt/postgresql/data/postgresql.conf`` to enable logging. Consider setting the ``log_directory`` to your preferred log file output location (here we use ``/home/<username>/logs`` where ``username`` is the name of your Uberspace).
```
Comment (brutus) [inline source/guide_postgresql.rst:237]:
```suggestion
Edit the "Reporting and Logging" section in ``~/opt/postgresql/data/postgresql.conf`` to enable logging. Consider setting the ``log_directory`` to your preferred log file output location (here we use ``/home/<username>/logs`` where ``username`` is the name of your Uberspace).
 
.. code-block:: postgres
  :emphasize-lines: 19
```

File: source/guide_postgresql.rst [hunk 5]
```diff
@@ -226,9 +231,56 @@ Consider using only unix sockets if possible.
...
+                                         # requires logging_collector to be on.
+
+ # This is used when logging to stderr:
+ logging_collector = on                  # Enable capturing of stderr and csvlog
+                                         # into log files. Required to be on for
+                                         # csvlogs.
+                                         # (change requires restart)
+
+ # These are only used if logging_collector is on:
+ log_directory = '/home/<username>/logs'                    # directory where log files are written,
```
Comment (brutus) [inline source/guide_postgresql.rst:257]:
```suggestion
 log_directory = '/home/<username>/logs'  # directory where log files are written,
```

File: source/guide_postgresql.rst [hunk 6]
```diff
@@ -226,9 +231,56 @@ Consider using only unix sockets if possible.
...
+                                         # happen after that much log output.
+                                         # 0 disables.
+
+
 
 Later you can see the socket in the filesystem by using ``ls -a ~/tmp``. It is listed as ``.s.PGSQL.5432``.
 
+
+
+
```
Comment (brutus) [inline source/guide_postgresql.rst:283]:
```suggestion
```

**Review by brutus — CHANGES_REQUESTED** [review]
@chhei I'll leave you some suggestons / questions.

File: source/guide_postgresql.rst [hunk 7]
```diff
@@ -298,14 +350,96 @@ Create Database
...
  [isabell@stardust ~]$
 
+PostgreSQL Extensions
+=====================
+
+PostgreSQL `extensions <https://wiki.postgresql.org/wiki/Extensions>`_ allow to extend the functionality and user-visible functions of a database. A number of extensions for PostgreSQL are available (see `list <https://www.postgresql.org/download/products/6-postgresql-extensions/>`_), with the geospatial extension PostGIS_ being one of the most widely used ones.
+
+Available extensions for PostgreSQL on U7 can be found in the ``/usr/pgsql-<MajorVersion>/share/extension`` directory, where ``<MajorVersion>`` refers to the PostgreSQL `Version`_ of your installation.
+
+.. Note:: In order to create extensions, the PostgreSQL user will need to have administrator rights. These are automatically assigned to the database owner using the ``--owner`` flag, as shown in the `Create database` section.
```
[the remark this answers is not in the document]
Comment (chhei) [reply source/guide_postgresql.rst]: @brutus - you were correct, I had created my db owner with the `--superuser` flag. Have corrected this in the guide and also linked to the PostgreSQL manual for more options on user and db creation.

File: source/guide_postgresql.rst (hunk 2 above)
Comment (chhei) [reply source/guide_postgresql.rst]: This is probably not needed anymore as PostgreSQL automatically uses the unix login name (in our case the Uberspace account name) when no `--username` is provided. For installing extensions this means that we're directly in superuser mode and can install extensions. This is clarified in the new version of the guide.

File: source/guide_postgresql.rst (hunk 3 above)
Comment (chhei) [reply source/guide_postgresql.rst]: see above comment

File: source/guide_postgresql.rst (hunk 4 above)
Comment (chhei) [reply source/guide_postgresql.rst]: fixed

File: source/guide_postgresql.rst (hunk 5 above)
Comment (chhei) [reply source/guide_postgresql.rst]: fixed

**Comment by brutus** [comment]
Sorry for the daily. This looks good, will try out and merge tomorrow.

**Comment by nichtmax** [comment]
@chhei Just a quick heads up: Janto is on vacation, is it okay if we review (and probably merge) your PR when he's back?

**Comment by chhei** [comment]
@nichtmax no worries - it's been hanging in here for a bit so a few more days won't likely make a huge difference.

**Comment by brutus** [comment]
Hi @chhei thanks again for the update (and your patience). Works fine now :+1:

**Comment by chhei** [comment]
Fantastic - great news!

**Pull request merged**